시간에 따른 코인 차트 그래프 데이터 갱신 기능 추가

donggeonoh opened the pull request:
- 코인 차트 그래프의 세그먼트 컨트롤 추가
- 세그먼트 컨트롤을 통한 그래프 데이터 갱신
- 그래프 최초 실행 시 최근 데이터 포지션으로 이동하는 기능 구현
- 상황에 맞는 이벤트를 처리하기 위해 DataManagerDelegate 메서드 추가
  - 최초로 그래프 데이터를 전달 할 때
  - 데이터를 추가 할 때
  - 데이터를 업데이트 할 때

- 이전 화면
![Mar-07-2022 14-16-54](https://user-images.githubusercontent.com/18098363/156972393-457399d8-4a17-4f16-af59-213e2b32cef5.gif)

- 현재 화면
<img width="358" alt="Screen Shot 2022-03-07 at 11 14 34 PM" src="https://user-images.githubusercontent.com/18098363/157050786-f7ac0967-99c6-4f6f-b497-8da15031c916.png">

donggeonoh committed:
feat: 시간에 따른 코인 차트 그래프 데이터 갱신 기능 추가 #59

- 코인 차트 그래프의 세그먼트 컨트롤 추가
- 세그먼트 컨트롤을 통한 그래프 데이터 갱신

z3rosmith commented on BithumbYagomAcamedy/Controller/CoinChartViewController.swift:
```diff
@@ -10,6 +10,7 @@ import Charts
 
 final class CoinChartViewController: UIViewController {
     @IBOutlet private weak var coinChartView: CandleStickChartView!
+    @IBOutlet weak var timeSegmentedControl: UISegmentedControl!
```
timeSegmentedControl가 private이 아닌데 CoinChartViewController 외에서는 쓰이지 않는것같네요!

z3rosmith commented on BithumbYagomAcamedy/Controller/CoinChartViewController.swift:
```diff
@@ -34,8 +35,21 @@ final class CoinChartViewController: UIViewController {
+        requestChartData()
+    }
+    
+    @IBAction func timeSegmentedControlValueChanged(_ sender: Any) {
+        requestChartData()
+    }
+    
+    private func requestChartData() {
+        guard let tickType = TickType(
+            rawValue: timeSegmentedControl.selectedSegmentIndex
+        ) else {
+            return
+        }
+        
+        dataManager?.changeTickType(to: tickType)
     }
```
`@IBOutlet timeSegmentedControl`를 사용하지 않고

```swift
@IBAction func timeSegmentedControlValueChanged(_ sender: UISegmentedControl) {
    requestChartData(segmentIndex: sender.selectedSegmentIndex)
}

private func requestChartData(segmentIndex: Int) {
    guard let tickType = TickType(
        rawValue: segmentIndex
    ) else {
        return
    }
    
    dataManager?.changeTickType(to: tickType)
}
```
이런 방식은 어떤가요??

donggeonoh replied:
아 그런 방식으로도 괜찮은 것 같네요. 의견 감사합니다 ㅎㅎ

donggeonoh replied:
최초 실행 시 timeSegmentedControl의 selectedIndex를 알아야해서 IBOutlet은 지우기 애매하네요.

donggeonoh committed:
feat: 피드백 내용 추가 및 spaceMax 설정 변경 # 59

donggeonoh committed:
feat: 그래프 최신 데이터 포지션으로 확대 기능 구현 및 delegate 추가

- 그래프의 첫 화면에서 최신 데이터 포지션으로 이동
- DataManagerDelegate를 추가하여 상황에 맞게(init, add, update) 데이터 전달

donggeonoh committed:
style: ChartDelegate 제거 및 print문 제거

HJEHA commented on BithumbYagomAcamedy/Controller/CoinChartViewController.swift:
```diff
@@ -45,14 +59,63 @@ extension CoinChartViewController: CoinChartDataManagerDelegate {
+    }
+    
+    private func moveChartPosition(entry: ChartDataEntry?) {
+        guard let x = entry?.x,
+              let y = entry?.y else {
+                  return
+              }
+        
```
컨벤션 확인 부탁드립니다:)

DonggeonOh replied:
혹시 어떤 부분의 컨벤션이 문제인걸까요?

donggeonoh replied:
넵 확인했습니다 ㅎㅎ

HJEHA commented on BithumbYagomAcamedy/Model/CoinChartDataManager.swift:
```diff
@@ -40,8 +38,29 @@ final class CoinChartDataManager {
+    private func requestChart() {
+        let api = CandlestickAPI(orderCurrency: symbol, tickType: tickType)
+        
+        httpService.request(api: api) { [weak self] result in
```
Result+extension 사용하시면 코드 뎁스를 줄이는데 도움이 될 것 같습니다:)

HJEHA commented on BithumbYagomAcamedy/Network/WebSocket/Model/TickType.swift:
```diff
@@ -1,29 +1,35 @@
     case minute30
     case hour1
     case hour24
     
```
웹 소켓 메세지 보낼 때 사용하려고 만든 타입인데 
CandleStick 차트에서 사용될 타입으로 변경된 것 같네요
바꾸신 이유를 알 수 있을까요??:)

기존에 가지고 있던 "30M", "1H", "12H", "24H", "MID" 케이스도 안보이는 것 같네요

donggeonoh replied:
이부분도 따로 타입 분리하여 처리하겠습니다

HJEHA reviewed (commented):
코멘트 남겼습니다:)
전체적인 컨벤션 확인 부탁드립니다:)

donggeonoh committed:
feat: 날짜 포맷 수정 및 TickType 역할 분리 #59

- 1분, 10분, 30분 날짜 포맷 yy.MM.dd\nhh:mm 으로 변경
- TickType을 ChartInterval, ChartDateFormat으로 각 역할에 맞게 세분화

HJEHA commented on BithumbYagomAcamedy/Model/Candlestick.swift:
```diff
@@ -41,20 +41,38 @@ struct Candlestick {
+        let endIndex = ticker.time.endIndex
+        var dateString: String
+        
+        switch dateFormat {
+        case .minute1, .minute10, .minute30:
+            let lastIndex = ticker.time.index(endIndex, offsetBy: -2)
+            formatter.dateFormat =  "yyyyMMddHHmm"
+            dateString = ticker.date + ticker.time[startIndex..<lastIndex]
+        case .hour1:
+            formatter.dateFormat =  "yyyyMMddHH"
+            let lastIndex = ticker.time.index(endIndex, offsetBy: -4)
+            dateString = ticker.date + ticker.time[startIndex..<lastIndex]
+        case .hour24:
+            formatter.dateFormat = "yyyyMMdd"
+            dateString = ticker.date
+        }
```
DateFormatter Extension를 사용하거나 별도의 타입을 생성해서 ChartDateFormat 타입을 매개변수로 받는 메서드로 분리시켜줘도 좋을 것 같습니다:)

donggeonoh replied:
넵 타입을 따로 분리해야겠네요 ㅎㅎ

donggeonoh committed:
feat: Candlestick init에 사용될 CandlestickDateFormatter 추가 #59

donggeonoh committed:
Merge branch 'develop' into feat/59/implement-interval-type-for-coin-chart-view-update
